web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
-            return HTTP_VALIDATION_NOT_SUPPORTED;
+            return NULL;
         }
 #endif
-
-        encoded_url = s = &s[7];
+        s = &s[7];
         w->mode = WEB_CLIENT_MODE_STREAM;
+    } else {
+        s = NULL;
     }
-    else {
+
+    return s;
+}
+
+/**
+ *  Find Protocol
+ *
+ *  The end of the url is the place where starts the protocol of the message,
+ *  this function finds the exact position where it starts.
+ *
+ * @param w is the structure with all information of the client request.
+ * @param s is the start of the user request
+ *
+ * @return
+ */
+static inline char *web_client_find_protocol(struct web_client *w,char *s) {
+    s = url_find_protocol(s);
+    if (s) {
+        //I must add 1, because I searched the separator+protocol
+        w->protocol.body = s+1;
+        //I know this string has at least ' HTTP/', so I do not need to search from the beginning
+        char *end = strchr(s+6,'\n');
+        if (end) {
+            w->protocol.length = end - w->protocol.body ;
+        }
+    }
+
+    return s;
+}
+
+/**
+ *  Parse header
+ *
+ *  Parse the headers sent together with the URL.
+ *
+ * @param w is the structure with all information of the client request.
+ * @param s is the string where the URL finished
+ */
+static inline void web_client_parse_headers(struct web_client *w,char *s) {
+    while(*s) {
+        // find a line feed
+        while(*s && *s++ != '\r');
+
+        // did we reach the end?
+        if(unlikely(!*s)) break;
+
+        if (*s == '\n') {
+            s++;
+        }
+
+        s = http_header_parse(w, s,
+                              (w->mode == WEB_CLIENT_MODE_STREAM) // parse user agent
+        );
+
+    }
+}
+
+/**
+ *  Parse request
+ *
+ * @param w is the structure with all information of the client request.
+ * @param divisor the place with the first equal sign
+ *
+ * @return It returns the number of variables in the query string.
+ */
+int web_client_parse_request(struct web_client *w,char *divisor) {
+    if(!divisor) {
+        w->total_params = 0;
+        return 0;
+    }
+
+    uint32_t i = url_parse_query_string(w->param_name,w->param_values,w->query_string.body+1,divisor);
+    w->total_params = i;
+
+    return i;
+}
+
+/**
+ *  Set Directory
+ *
+ *  Map directory variables
+ *
+ * @param w is the structure with all information of the client request.
+ * @param begin is the end of the first directory
+ * @param enddir a pointer to the position with the last valid character
+ * @param endcmd the end of the last character.
+ */
+static inline void web_client_set_directory(struct web_client *w,char *begin,char *enddir,char *endcmd) {
+    if (enddir) {
+        w->directory.body = begin;
+        w->directory.length = enddir - begin;
+
+        if (!strncmp(w->directory.body,"api",w->directory.length)) {
+            begin = enddir + 1;
```

Comment:
Logic: can this get out of bounds? Is `enddir + 1` always smaller than `endcmd`?

Reply:
Yes it is. The "enddir" is a pointer for the PATH in the first line of the request, and the request structure in the first line will always have structure like this:

METHOD PATH HTTP/1.1\r\n\r\n

So we will have at least 13 bytes after the path.